From e76245a078bd8f946ce8d667ce2e3b6ec4acb7e8 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Wed, 8 Dec 2021 14:29:22 +0100
Subject: [PATCH 0528/1695] plat-stm32mp1: SiP SMC services for low power

SiP and OEM SMC services for low power support, related to configuration
CFG_STM32_LOWPOWER_SIP=y which depends on CFG_PM=y.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f44207f7dda2cad4ba40dd71981da4e1728a40cf plat-stm32mp1: SiP SMC services for low power]
Change-Id: Id9b98ad28e4d6f90594d1814ed7d58e01941e6d8
Signed-off-by: Etienne Carriere <etienne.carriere@st.com>
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/358374
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
---
 core/arch/arm/plat-stm32mp1/conf.mk           |  3 +-
 .../nsec-service/low_power_svc.c              | 37 +++++++++++++++++++
 .../nsec-service/low_power_svc.h              | 23 ++++++++++++
 .../arm/plat-stm32mp1/nsec-service/pwr_svc.c  |  1 +
 .../plat-stm32mp1/nsec-service/stm32mp1_smc.h | 18 ++++++++-
 .../nsec-service/stm32mp1_svc_setup.c         |  4 ++
 .../arm/plat-stm32mp1/nsec-service/sub.mk     |  1 +
 7 files changed, 85 insertions(+), 2 deletions(-)
 create mode 100644 core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.c
 create mode 100644 core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.h

diff --git a/core/arch/arm/plat-stm32mp1/conf.mk b/core/arch/arm/plat-stm32mp1/conf.mk
index 459e5d07a..8120e4d0e 100644
--- a/core/arch/arm/plat-stm32mp1/conf.mk
+++ b/core/arch/arm/plat-stm32mp1/conf.mk
@@ -303,7 +303,8 @@ endif
 # SiP/OEM service for non-secure world
 # BSEC fuse access, PWR services, SCMI transport.
 CFG_STM32_BSEC_SIP ?= n
-CFG_STM32_PWR_SIP ?= $(CFG_STM32MP15)
+CFG_STM32_LOWPOWER_SIP ?= y
+CFG_STM32_PWR_SIP ?= y
 CFG_STM32MP1_SCMI_SIP ?= n
 ifeq ($(CFG_STM32MP1_SCMI_SIP),y)
 $(call force,CFG_SCMI_MSG_DRIVERS,y,Mandated by CFG_STM32MP1_SCMI_SIP)
diff --git a/core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.c b/core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.c
new file mode 100644
index 000000000..e2e9ad766
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.c
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2017-2018, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2017, ARM Limited and Contributors. All rights reserved.
+ */
+
+#include <drivers/stm32mp_dt_bindings.h>
+#include <sm/pm.h>
+#include <sm/psci.h>
+#include <stm32mp_pm.h>
+#include <stm32_util.h>
+#include <trace.h>
+
+#include "stm32mp1_smc.h"
+#include "low_power_svc.h"
+#include "../pm/power.h"
+#include "../pm/context.h"
+
+uint32_t pm_domain_scv_handler(uint32_t id, uint32_t enable)
+{
+	unsigned int pd = id;
+
+	DMSG("%sable PD %u", enable != 0 ? "En" : "Dis", pd);
+
+	switch (pd) {
+	case STM32MP1_PD_VSW:
+	case STM32MP1_PD_CORE_RET:
+	case STM32MP1_PD_CORE:
+		break;
+	default:
+		return STM32_SIP_SVC_INVALID_PARAMS;
+	}
+
+	stm32mp1_set_pm_domain_state(pd, enable);
+
+	return STM32_SIP_SVC_OK;
+}
diff --git a/core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.h b/core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.h
new file mode 100644
index 000000000..520a26846
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp1/nsec-service/low_power_svc.h
@@ -0,0 +1,23 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2018-2021, STMicroelectronics
+ */
+
+#ifndef LOW_POWER_SVC_H
+#define LOW_POWER_SVC_H
+
+#include <sm/sm.h>
+#include <stdint.h>
+
+#include "stm32mp1_smc.h"
+
+#ifdef CFG_STM32_LOWPOWER_SIP
+uint32_t pm_domain_scv_handler(uint32_t x1, uint32_t x2);
+#else
+static inline uint32_t pm_domain_scv_handler(uint32_t x1 __unused,
+					     uint32_t x2 __unused)
+{
+	return STM32_SIP_SVC_FAILED;
+}
+#endif /* CFG_STM32_LOWPOWER_SIP */
+#endif /* LOW_POWER_SVC_H */
diff --git a/core/arch/arm/plat-stm32mp1/nsec-service/pwr_svc.c b/core/arch/arm/plat-stm32mp1/nsec-service/pwr_svc.c
index 1c3fd7c9f..9aa65c01b 100644
--- a/core/arch/arm/plat-stm32mp1/nsec-service/pwr_svc.c
+++ b/core/arch/arm/plat-stm32mp1/nsec-service/pwr_svc.c
@@ -6,6 +6,7 @@
 #include <inttypes.h>
 #include <drivers/stm32_shared_io.h>
 #include <drivers/stm32mp1_pwr.h>
+#include <drivers/stm32_shared_io.h>
 #include <kernel/panic.h>
 #include <io.h>
 #include <mm/core_memprot.h>
diff --git a/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h b/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h
index 8a0a6cb90..803e7a0e9 100644
--- a/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h
+++ b/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_smc.h
@@ -14,7 +14,7 @@
 #define STM32_SIP_SVC_VERSION_MAJOR	0x0
 #define STM32_SIP_SVC_VERSION_MINOR	0x1
 
-#define STM32_SIP_SVC_FUNCTION_COUNT	0x4
+#define STM32_SIP_SVC_FUNCTION_COUNT	0x5
 
 /* STM32 SIP service generic return codes */
 #define STM32_SIP_SVC_OK		0x0
@@ -110,6 +110,22 @@
 /* reserved for STM32_SIP_SVC_SMC_WRITE_ALL	0x6 */
 #define STM32_SIP_SVC_BSEC_WRLOCK_OTP		0x7
 
+/*
+ * SIP function STM32_SIP_FUNC_PD_DOMAIN
+ *
+ * Argument a0: (input) SMCC ID
+ *		(output) status return code
+ * Argument a2: (index) ID of target power domain to be enabled/disabled
+ * Argument a3: (input) 0 to disable, 1 to enable target domain
+ */
+#define STM32_SIP_FUNC_PD_DOMAIN		0x1008
+
+/* Valid IDs for power domain for function STM32_SIP_FUNC_PD_DOMAIN */
+#define STM32_SIP_PD_VSW			0x0
+#define STM32_SIP_PD_CORE_RET			0x1
+#define STM32_SIP_PD_CORE			0x2
+#define STM32_SIP_PD_MAX_PM_DOMAIN		0x3
+
 /*
  * SIP function STM32_SIP_SVC_FUNC_SCMI_AGENT
  * SIP function STM32_SIP_SVC_FUNC_SCMI_AGENT1 - Obsolete
diff --git a/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c b/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c
index 27c0a74bb..db656563c 100644
--- a/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c
+++ b/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c
@@ -11,6 +11,7 @@
 #include <sm/std_smc.h>
 
 #include "bsec_svc.h"
+#include "low_power_svc.h"
 #include "pwr_svc.h"
 #include "stm32mp1_smc.h"
 
@@ -48,6 +49,9 @@ static enum sm_handler_ret sip_service(struct sm_ctx *ctx __unused,
 	case STM32_SIP_SVC_FUNC_PWR:
 		args->a0 = pwr_scv_handler(args->a1, args->a2, args->a3);
 		break;
+	case STM32_SIP_FUNC_PD_DOMAIN:
+		args->a0 = pm_domain_scv_handler(args->a1, args->a2);
+		break;
 	default:
 		return SM_HANDLER_PENDING_SMC;
 	}
diff --git a/core/arch/arm/plat-stm32mp1/nsec-service/sub.mk b/core/arch/arm/plat-stm32mp1/nsec-service/sub.mk
index 25dd50251..f297caafe 100644
--- a/core/arch/arm/plat-stm32mp1/nsec-service/sub.mk
+++ b/core/arch/arm/plat-stm32mp1/nsec-service/sub.mk
@@ -2,4 +2,5 @@ global-incdirs-y += .
 
 srcs-y += stm32mp1_svc_setup.c
 srcs-$(CFG_STM32_BSEC_SIP) += bsec_svc.c
+srcs-$(CFG_STM32_LOWPOWER_SIP) += low_power_svc.c
 srcs-$(CFG_STM32_PWR_SIP) += pwr_svc.c
-- 
2.47.2

